pkgs/stdenv/booter.nix

```diff
@@ -0,0 +1,69 @@
+{ lib, allPackages }:
+
+rec {
+  # It would be nice to just use `++`, but we can't forget about the old
+  # `basecase`. I (@Ericson2314) am tempted to use a heterogeneous list instead.
+  extendStages = { baseCase, stageFuns }: more: {
+    inherit baseCase;
+    stageFuns = stageFuns ++ more;
+  };
+
+  # Type:
+  #   {
+  #     baseCase : pkgSet
+  #     stageFuns : [ pkgset -> args to stage/default.nix ]
+  #   }
+  #   -> pkgset
+  #
+  # In english, this takes:
+  #
+  #  - a list of function from the previous stage pkgset, to args for this stage
+  #  - pkgset (the most complex and important arg is the stdenv)
+  #
+  #  - An initial pkg set, usually from a hardcoded bootstrap tools tarball
+  #
+  #  and returns the final pkgset.
+  #
+  # The list takes stages in order, so the final stage is last in the list. In
+  # other words, this does a foldr not foldl.
+  boot = { baseCase, stageFuns }: let
+
+    # Take the list and disallow custom overrides in all but the final stage,
+    # and allow it in the final flag. Only defaults this boolean field if it
+    # isn't already set.
+    withAllowCustomOverrides = lib.lists.imap
+      (index: stageFun: prevStage:
+        { allowCustomOverrides = index == 1; } # first element, 1-indexed
+        // (stageFun prevStage))
+      stageFuns;
+
+    # Adds the stdenv to the arguments, and sticks in it the previous stage for
+    # debugging purposes.
+    folder = finalSoFar_: stageFun: let
+        # Build wrappers are always added to the previous stage, and always
+        # always added to build packages. That means we need to also extend the
+        # current stage if it is used as its own build packages. Note that the
+        # *returned* current stage is never so extended, regardless of what
+        # build packages are used.
+        buildWrappers = import ../top-level/build-wrappers.nix self;
+        finalSoFar = finalSoFar_.overridePackages buildWrappers;
+
+        args_ = stageFun finalSoFar;
+        selfBuild = args_.selfBuild or true;
+        args = lib.attrsets.filterAttrs (k: _: k != "selfBuild") args_;
+        inherit (args) stdenv;
+        self = allPackages (args // {
+          buildPackages =
+            if selfBuild
+            then self.overridePackages buildWrappers
```

Comment:
I tried to explain in the comments above, but could probably be clearer :). I only want `buildPackages` to have the build wrappers, as those wrappers are not needed at runtime, ever. Due to laziness, only one `overridePackages` call is ever evaluated.